### PR TITLE
Cleanup TP

### PR DESCRIPTION
**Pull request by eitanturok**
This PR cleans up TP:
* Removes `@experimental_function('Tensor Parallelism')` from `build_tp_stratigies` because we already raise this warning in composer's TP [implementation](https://github.com/mosaicml/composer/blob/e5e2f744baa9f5bec8ff24b58f5a6b2a564919d0/composer/core/state.py#L600).
* Don't do a deep-copy of tp_config, instead create a new tp_config in `train.py`. This makes things a bit cleaner.
* Check that `tp_config` contains both 'strategy' and 'tensor_parallel_degree' fields.
* We previously only checked the tp edge cases but never had a simple example that tested tp actually working. So I added the ` test_tp_train` test to do this.
* I gave the tests better names

### CODE REVIEW

File: llmfoundry/utils/builders.py [hunk 1]
```diff
@@ -701,7 +701,6 @@ def _validate_cfg(icl_cfg: dict[str, Any]):
     return evaluators, logger_keys
 
 
-@experimental_function('Tensor Parallelism')
```
Comment (dakinggg) [inline llmfoundry/utils/builders.py:704]: I think we should leave this personally. We may want to remove the experimental designation in Composer and Foundry at different times.

File: llmfoundry/command_utils/train.py [hunk 2]
```diff
@@ -518,8 +517,11 @@ def train(cfg: DictConfig) -> Trainer:
     # TP config
     if tp_config is not None:
         strategy = tp_config.pop('strategy', None)
-        assert isinstance(strategy, str), '`strategy` must be in `tp_config`.'
-        tp_config['layer_plan'] = build_tp_strategies(strategy, model)
+        layer_plan = build_tp_strategies(strategy, model)
+        tp_config = {
```
Comment (dakinggg) [inline llmfoundry/command_utils/train.py:521]: the compose tp config has other keys allowed. should we just be constructing that dataclass here?

File: tests/tp/test_tp_strategies.py [hunk 3]
```diff
@@ -97,8 +99,34 @@ def test_ffn_tp_strategy():
 
 
 @pytest.mark.gpu
-def test_no_tp_with_one_gpu():
-    """Test that when we have one GPU, we use DDP and not FSDP-TP."""
+@pytest.mark.world_size(4)
+@pytest.mark.parametrize('tp_strategy', ['ffn'])
+def test_tp_train(tp_strategy: str):
```
Comment (dakinggg) [inline tests/tp/test_tp_strategies.py:104]: tests like this are reasonably costly and also can be hard to debug. is there a simpler test that would achieve the same coverage?

File: llmfoundry/utils/builders.py (hunk 1 above)
Comment (mvpatel2000) [reply llmfoundry/utils/builders.py:704]: +1

File: llmfoundry/command_utils/train.py (hunk 2 above)
Comment (mvpatel2000) [reply llmfoundry/command_utils/train.py:521]: +1

File: llmfoundry/utils/builders.py (hunk 1 above)
Comment (eitanturok) [reply llmfoundry/utils/builders.py]: added back

File: llmfoundry/command_utils/train.py (hunk 2 above)
Comment (eitanturok) [reply llmfoundry/command_utils/train.py]: done

File: tests/tp/test_tp_strategies.py (hunk 3 above)
Comment (eitanturok) [reply tests/tp/test_tp_strategies.py:104]: I believe this test is okay because
* we only train for one step
* we want to make sure things work end-to-end
* 
I spoke to @mvpatel2000 about this and he said he thought it was okay.

File: llmfoundry/command_utils/train.py [hunk 4]
```diff
@@ -517,9 +516,9 @@ def train(cfg: DictConfig) -> Trainer:
 
     # TP config
     if tp_config is not None:
-        strategy = tp_config.pop('strategy', None)
-        assert isinstance(strategy, str), '`strategy` must be in `tp_config`.'
-        tp_config['layer_plan'] = build_tp_strategies(strategy, model)
+        strategy = tp_config.pop('strategy')
+        layer_plan = build_tp_strategies(strategy, model)
```
Comment (snarayan21) [inline llmfoundry/command_utils/train.py:520]: is dataloading replication not needed in this PR? or is it already in

Comment (eitanturok) [reply llmfoundry/command_utils/train.py:520]: discussed offline.
